src/transport/SecureSession.cpp

```diff
@@ -144,12 +151,18 @@ void SecureSession::MarkInactive()
     case State::kDefunct:
         FALLTHROUGH;
     case State::kActive:
+        // Make sure we don't notify "deferred" that we're being removed.  This
+        // ensures that "deferred" is in fact in our holder list.
+        RemoveHolder(deferred);
         // By setting this state, IsActiveSession() will return false, which prevents creating new exchanges.
-        mState = State::kInactive;
+        MoveToState(State::kInactive);
+        NotifySessionReleased();
```

Comment:
Why is this necessary? When a session is marked for release, it's not strictly necessary that you detach all existing holders right? That can happen later when the session _actually_ gets marked for eviction no?

As long as we prevent new exchanges from being created on the in-active session, that's what matters I'd think...

Reply:
We need to detach all existing exchanges (except our special one) that happen to be using this session already.

We can either do that via notifying them the session is gone, or by telling the exchanges something directly.  This does the former, because that's a tested codepath that we use elsewhere anyway.  We could do that latter, but it means more codepaths to maintain, test, and reason about...

Reply:
Yes, per spec those exchanges need to die immediately.  e.g. if we're in the middle of a chunked read, it needs to stop delivering data, no?

Reply:
Does the spec actually say exchanges need to die immediately?

> > All internal data reflecting the prior operational identifier of the Node within the Fabric SHALL be revoked and removed, to an outcome equivalent to the disappearance of the prior Node, except for the ongoing CASE session context, which SHALL temporarily remain valid until the NOCResponse has been successfully delivered or until the next transport-layer error, so that the response can be received by the Administrator invoking the command.

I guess it kinda does...